src/Symfony/Component/HttpClient/RetryableHttpClient.php

```diff
@@ -54,40 +55,72 @@ public function __construct(HttpClientInterface $client, RetryDeciderInterface $
     public function request(string $method, string $url, array $options = []): ResponseInterface
     {
         $retryCount = 0;
+        $content = '';
+        $firstChunk = null;
 
-        return new AsyncResponse($this->client, $method, $url, $options, function (ChunkInterface $chunk, AsyncContext $context) use ($method, $url, $options, &$retryCount) {
+        return new AsyncResponse($this->client, $method, $url, $options, function (ChunkInterface $chunk, AsyncContext $context) use ($method, $url, $options, &$retryCount, &$content, &$firstChunk) {
             $exception = null;
             try {
                 if ($chunk->isTimeout() || null !== $chunk->getInformationalStatus()) {
                     yield $chunk;
 
                     return;
                 }
-
-                // only retry first chunk
-                if (!$chunk->isFirst()) {
-                    $context->passthru();
-                    yield $chunk;
-
-                    return;
-                }
             } catch (TransportExceptionInterface $exception) {
                 // catch TransportExceptionInterface to send it to strategy.
             }
 
-            $statusCode = $context->getStatusCode();
-            $headers = $context->getHeaders();
-            if ($retryCount >= $this->maxRetries || !$this->decider->shouldRetry($method, $url, $options, $partialResponse = new MockResponse($context->getContent(), ['http_code' => $statusCode, 'headers' => $headers]), $exception)) {
+            if ($retryCount >= $this->maxRetries) {
                 $context->passthru();
                 yield $chunk;
 
                 return;
             }
 
+            $statusCode = $context->getStatusCode();
+            $headers = $context->getHeaders();
+            if ($chunk instanceof FirstChunk || null !== $exception) {
+                $shouldRetry = $this->decider->shouldRetry($method, $url, $options, $context->getStatusCode(), $context->getHeaders(), null, $exception);
+
+                if (false === $shouldRetry) {
+                    $context->passthru();
+                    yield $chunk;
+
+                    return;
+                }
+
+                // Decider need body to decide
+                if (null === $shouldRetry) {
+                    $firstChunk = $chunk;
+                    $content = '';
+                    // disable exception thrown on destruct
+                    $context->getResponse()->getStatusCode();
```

Comment:
do we really need this?

Reply:
If not, and the response return a non-200 code, we get:

```
PHP Fatal error:  Uncaught Symfony\Component\HttpClient\Exception\ClientException: HTTP/1.1 404 Not Found returned for "https://github.com/jderusse404". in /data/oss/symfony/src/Symfony/Component/HttpClient/Response/CommonResponseTrait.php:176
Stack trace:
#0 /data/oss/symfony/src/Symfony/Component/HttpClient/Response/TransportResponseTrait.php(68): Symfony\Component\HttpClient\Response\CurlResponse->checkStatusCode()                                                
#1 /data/oss/symfony/src/Symfony/Component/HttpClient/Response/TransportResponseTrait.php(247): Symfony\Component\HttpClient\Response\CurlResponse->getHeaders(true)                                                
#2 [internal function]: Symfony\Component\HttpClient\Response\CurlResponse::stream(Array, NULL)
#3 /data/oss/symfony/src/Symfony/Component/HttpClient/Response/ResponseStream.php(42): Generator->next()
#4 /data/oss/symfony/src/Symfony/Component/HttpClient/Response/AsyncResponse.php(206): Symfony\Component\HttpClient\Response\ResponseStream->next()                                                                 
#5 /data/oss/symfony/src/Symfony/Component/HttpClient/Response/CommonResponseTrait.php(151): Symfony\Compo in /data/oss/symfony/src/Symfony/Component/HttpClient/Response/CommonResponseTrait.php on line 176       
```

Reply:
I feel like we should move this call to `AsyncResponse` and passthru filter shouldn't have to deal with it